dual/dual.go

```diff
@@ -153,31 +154,60 @@ func (dht *DHT) FindPeer(ctx context.Context, pid peer.ID) (peer.AddrInfo, error
 
 	wg.Wait()
 
-	// combine addresses
-	deduped := make(map[string]ma.Multiaddr, len(wanInfo.Addrs)+len(lanInfo.Addrs))
-	for _, addr := range wanInfo.Addrs {
-		deduped[string(addr.Bytes())] = addr
+	// Combine addresses. Try to avoid doing unnecessary work while we're at
+	// it. Note: We're ignoring the errors for now as many of our DHT
+	// commands can return both a result and an error.
+	ai := peer.AddrInfo{ID: pid}
+	if len(wanInfo.Addrs) == 0 {
+		ai.Addrs = lanInfo.Addrs
+	} else if len(lanInfo.Addrs) == 0 {
+		ai.Addrs = wanInfo.Addrs
+	} else {
+		// combine addresses
+		deduped := make(map[string]ma.Multiaddr, len(wanInfo.Addrs)+len(lanInfo.Addrs))
+		for _, addr := range wanInfo.Addrs {
+			deduped[string(addr.Bytes())] = addr
+		}
+		for _, addr := range lanInfo.Addrs {
+			deduped[string(addr.Bytes())] = addr
+		}
+		ai.Addrs = make([]ma.Multiaddr, 0, len(deduped))
+		for _, addr := range deduped {
+			ai.Addrs = append(ai.Addrs, addr)
+		}
 	}
-	for _, addr := range lanInfo.Addrs {
-		deduped[string(addr.Bytes())] = addr
+
+	// If one of the commands succeeded, don't return an error.
+	if wanErr == nil || lanErr == nil {
+		return ai, nil
 	}
-	addrs := make([]ma.Multiaddr, 0, len(deduped))
-	for _, addr := range deduped {
-		addrs = append(addrs, addr)
+
+	// Otherwise, return what we have _and_ return the error.
+	return ai, combineErrors(wanErr, lanErr)
+}
+
+func combineErrors(erra, errb error) error {
+	// if the errors are the same, just return one.
+	if erra == errb {
```

Comment:
can 2 instances of the same error not be `==`? will this do string equality (which should be safe)?

Reply:
This won't do string equality, but I really just care if they're the same. E.g., "ErrNotFound", "ErrNotSupported", etc. Unfortunately, this is just a heuristic to improve the errors a bit.